### DISCUSSION TITLE
Codegen runs even in case of old arch.

### DISCUSSION THREAD

**Discussion by varunkukade** [callstack/react-native-builder-bob#745] [General]
I created a library using create-react-native-library. It works great on projects with new arch enabled. But when I try to run it on projects with old arch, codegen generates the code in generated folder and hence XCode builds fails with "/ios/generated/RNPackageNameSpecJSI.h:12:10 'ReactCommon/TurboModule.h' file not found. 

Is there anything I am missing. This is my first time creating library.

@atlj

**Comment by satya164** [comment]
What options did you select when creating the library? Does this change make a difference? https://github.com/callstack/react-native-builder-bob/pull/737/files

**Comment by varunkukade** [comment]
I selected "Turbo Module" while creating the library. I tried to run it on react native version 0.72.4 which is an old/legacy architecture-based project.

I am not able to patch this change https://github.com/callstack/react-native-builder-bob/pull/737/files into the react-native-builder-bob library. Should I directly put this in my library's podspec?

Also, this change seems to be related to new architecture.

**Reply by satya164** [reply]
Turbo modules are part of new architecture. You can't use a turbo module with old architecture. Use a legacy native module instead.

> Also, this change seems to be related to new architecture.

So is codegen.

**Reply by Alexufo** [reply]
There's a flag inside the gradle isNewArchitecture. What is it for?

**Comment by varunkukade** [comment]
I followed these steps
1. Created library using create-react-native-library. 
2. I added all these steps listed here to support old arch. https://github.com/reactwg/react-native-new-architecture/blob/main/docs/backwards-compat-turbo-modules.md. According to this I used the imports and implementation based on RCT_NEW_ARCH_ENABLED. This way it used legacy module in case of old arch.
3. I ran yarn prepare - which compiles the code using Babel, and generates the code using codegen. 
4. Now I published this package and used it in the old arch-based project. I tried to build the ios project. It gave me this issue. 
`"/ios/generated/RNPackageNameSpecJSI.h:12:10 'ReactCommon/TurboModule.h' file not found."`
5. This issue came from the generated folder that codegen generates while doing yarn prepare. 

hence I removed the **generated** ios folder and then tried to use this library as local package into my old arch project.
And error went away. I was able to build the project successfully as I don't have a generated folder. As the **generated** folder is included by default when you install the package in old arch-based project, I got the error.

Now I see that we have **includesGeneratedCode** as codegen config which includes the generated folder by default with package install. If I make it false and also remove **outputDir**, while yarn prepare I get issue as 
`✖ Cannot read properties of undefined (reading 'android')`. 

Is this same issue here https://github.com/callstack/react-native-builder-bob/issues/713 ? If yes, how can I apply this patch? https://github.com/callstack/react-native-builder-bob/pull/716

Am I in right direction?

**Reply by varunkukade** [reply]
Also even if I install the library in the old arch project and manually go to node_modules and delete the **generated** folder for IOS, this error goes away. Is there any solution to not generate **"generated"** folder by codegen conditionally based on the consumer project's arch?

**Reply by vkukade-altir** [reply]
@satya164 is there something I am missing here?

**Reply by satya164** [reply]
I currently don't have the bandwidth to look into it. However, codegen in the app is run by the app - and those parts are maintained by React Native team. So I don't think I'd be able to provide any insights there.

**Reply by Alexufo** [reply]
@varunkukade How you organize oldarh in grandle? Can you share this part?
Documentation recommend do this
````
    sourceSets {
        main {
            if (isNewArchitectureEnabled()) {
                java.srcDirs += ['src/newarch']
            } else {
                java.srcDirs += ['src/oldarch']
            }
     }
````     

but bob builder has this
```
    main {
          java.srcDirs += [
            "generated/java",
            "generated/jni"
          ]
```

**Reply by vkukade-altir** [reply]
@satya164 thanks for your input. I was able to solve this issue with some workaround. 

But here is another discussion thread which includes correct solution by Meta Team. https://github.com/reactwg/react-native-new-architecture/discussions/8#discussioncomment-11703117

**Reply by varunkukade** [reply]
@Alexufo You can refer to my commits in this repo. I was able to successfully do this.
Check commits on Jan 1, 2025 and Jan 4, 2025 on this repo https://github.com/varunkukade/react-native-skaleton-kit/commits/main/

**Reply by Alexufo** [reply]
@varunkukade thanks. 
You didn't change anything on the js side? Native module is linked the same way as turbo module?
https://github.com/reactwg/react-native-new-architecture/blob/main/docs/backwards-compat-turbo-modules.md#unify-the-javascript-specs
